beacon_node/beacon_chain/src/attestation_rewards.rs

```diff
@@ -0,0 +1,191 @@
+                if !state.is_in_inactivity_leak(previous_epoch, spec) {
+                    ideal_rewards_hashmap
+                        .insert((flag_index, effective_balance_eth), (ideal_reward, penalty));
+                } else {
+                    ideal_rewards_hashmap.insert((flag_index, effective_balance_eth), (0, penalty));
+                }
+            }
+        }
+
+        // Calculate total_rewards
+        let mut total_rewards: Vec<TotalAttestationRewards> = Vec::new();
+
+        if validators.is_empty() {
+            validators = participation_cache.eligible_validator_indices().to_vec();
+        }
+
+        for validator_index in &validators {
+            let eligible = state.is_eligible_validator(previous_epoch, *validator_index)?;
+            let mut head_reward = 0u64;
+            let mut target_reward = 0i64;
+            let mut source_reward = 0i64;
+
+            if eligible {
+                let effective_balance = state.get_effective_balance(*validator_index)?;
+
+                let effective_balance_eth =
+                    effective_balance.safe_div(spec.effective_balance_increment)?;
+
+                for flag_index in 0..PARTICIPATION_FLAG_WEIGHTS.len() {
+                    let (ideal_reward, penalty) = ideal_rewards_hashmap
+                        .get(&(flag_index, effective_balance_eth))
+                        .ok_or(BeaconChainError::AttestationRewardsError)?;
+                    let voted_correctly = participation_cache
+                        .get_unslashed_participating_indices(flag_index, previous_epoch)
+                        .map_err(|_| BeaconChainError::AttestationRewardsError)?
+                        .contains(*validator_index)
+                        .map_err(|_| BeaconChainError::AttestationRewardsError)?;
+                    if voted_correctly {
+                        if flag_index == TIMELY_HEAD_FLAG_INDEX {
+                            head_reward += ideal_reward;
+                        } else if flag_index == TIMELY_TARGET_FLAG_INDEX {
+                            target_reward += *ideal_reward as i64;
+                        } else if flag_index == TIMELY_SOURCE_FLAG_INDEX {
+                            source_reward += *ideal_reward as i64;
+                        }
+                    } else if flag_index == TIMELY_HEAD_FLAG_INDEX {
+                        head_reward = 0;
+                    } else if flag_index == TIMELY_TARGET_FLAG_INDEX {
+                        target_reward = *penalty;
+                    } else if flag_index == TIMELY_SOURCE_FLAG_INDEX {
+                        source_reward = *penalty;
+                    }
+                }
+            } else {
+                total_rewards.push(TotalAttestationRewards {
+                    validator_index: *validator_index as u64,
+                    head: 0,
+                    target: 0,
+                    source: 0,
+                });
+            }
```

Comment:
I don't think this branch is necessary. I see you added it in 7c09a6d2ee0c3b46d167926c1eb4eacdd3c7584c to _prevent_ duplicate responses but it actually has the opposite effect. When the requested validator is ineligible (e.g. 200075) it will result in two lots of rewards:

```
$ curl --data "[20075]" -H "Content-Type:application/json" "http://localhost:5052/eth/v1/beacon/rewards/attestations/176130" | jq
```

```json
{
  "execution_optimistic": true,
  "data": {
    "ideal_rewards": [], // ommitted
    "total_rewards": [
      {
        "validator_index": "20075",
        "head": "0",
        "target": 0,
        "source": 0
      },
      {
        "validator_index": "20075",
        "head": "0",
        "target": 0,
        "source": 0
      }
    ]
  }
}

```